pkg/promise/promise.go

```diff
@@ -49,16 +65,47 @@ func (p *promise) Complete(errors []error) {
 	}
 }
 
-// Blocks the caller until the promise is marked complete.
+// Blocks the caller until the promise is marked complete. This function
+// is equivalent to invoking AwaitUntil() with a zero-length duration.
+// To avoid blocking the caller indefinitely, use AwaitUntil() with a
+// non-zero duration instead.
 func (p *promise) Await() []error {
-	<-p.completeChan
-	return p.errors
+	return p.AwaitUntil(0 * time.Second)
 }
 
-// Invokes the supplied function after this promise completes.
+// Blocks the caller until the promise is marked complete, or the supplied
+// duration has elapsed. If the promise has not been completed before the
+// await times out, this function returns with nonempty errors. If the
+// supplied duration has zero length, this await will never time out.
+func (p *promise) AwaitUntil(duration time.Duration) []error {
+	var timeoutChan <-chan time.Time
+	if duration.Nanoseconds() > 0 {
+		timeoutChan = time.After(duration)
+	}
+
+	select {
+	case <-p.completeChan:
+		return p.errors
+	case <-timeoutChan:
+		msg := fmt.Sprintf("Await timed out for promise after [%s]", duration)
+		return []error{NewTimeoutError(msg)}
```

Comment:
I can make that change.  Admittedly, the custom error type doesn't add much in this case.